k2/csrc/array.h

```diff
@@ -49,73 +53,154 @@ struct Array1 {
 template <typename I>
 struct Array2Size {
   using IndexT = I;
-  // `size1` is the top-level size of the array, equal to the object's .size
+  // `size1` is the top-level size of the array, equal to the object's .size1
   // element
   I size1;
-  // `size2` is the nunber of elements in the array, equal to
-  // o->indexes[o->size] - o->indexes[0] (if the Array2 object o is
-  // initialized).
+  // `size2` is the number of elements in the array, equal to the object's
+  // .size2 element
   I size2;
 };
 
-template <typename I, typename Ptr>
+template <typename Ptr, typename I = int32_t>
 struct Array2 {
   // Irregular two dimensional array of something, like vector<vector<X> >
   // where Ptr is, or behaves like, X*.
   using IndexT = I;
   using PtrT = Ptr;
-
-  IndexT size;
-  const IndexT *indexes;  // indexes[0,1,...size] should be defined; note, this
-                          // means the array must be of at least size+1.  We
-                          // require that indexes[i] <= indexes[i+1], but it is
-                          // not required that indexes[0] == 0, it may be
-                          // greater than 0.
-
+  using ValueType = typename std::iterator_traits<Ptr>::value_type;
+
+  IndexT size1;
+  IndexT *indexes;  // indexes[0,1,...size1] should be defined; note, this
```

Comment:
It'd be better if the definition order of `indexes` and `size2` is swapped.

On 64-bit machines where `sizeof(indexes) == 8`, there will be 4-byte padding
after `size1` for the current implementation.